### DISCUSSION TITLE
Add new language server capabilities

### DISCUSSION THREAD

**Discussion by niontrix** [emacs-lsp/lsp-mode#4330] [General]
Hi folks,
there was a new Nimlangserver release recently, which adds some Nim specific capabilities (e.g. macroExpand) that obviously are not part of the default language server protocol. My question now is: How can I make  use of those? Do I need to create an lsp-mode plugin?
I hope someone can give me a hint. Maybe I have overlooked something in the documentation?!

Anyway,..
Thanks for making such a useful Emacs package and keep up the great work!

KR

**Comment by jcs090218** [comment]
IMO, it depends on the scale of the package. If the scale is small, you can simply contribute to this repo, see the [clients](https://github.com/emacs-lsp/lsp-mode/tree/master/clients) folder (see #4283 for reference). If you are planning to support more (medium to large scale), I prefer you create a separate package for this ([lsp-dart](https://github.com/emacs-lsp/lsp-dart), [lsp-java](https://github.com/emacs-lsp/lsp-java), etc).

**Reply by niontrix** [reply]
Currently the deviation from the default LSP are small, as far as I can tell, but I also don't know how much difference there will be in the future.

**Reply by jcs090218** [reply]
I prefer start in a separate package, then move into `lsp-mode` if needed. Like [lsp-powershell](https://github.com/kiennq/lsp-powershell) was started as a separated package, but moved into core after a while. Here are some considerations to ponder:

1. You don't have full control in this repo, development can be slow since you will need someone to review your PRs
2. From the user perspective, splitting up a package may be unwelcome experiences. But move into the core don't.

Hope the information helps! :)

**Comment by yyoncho** [comment]
@niontrix I used to work on nimlangserver and I have this: https://github.com/yyoncho/lsp-mode/commit/b9ad0573a50743149714f57b16377dc3aef0e24b . Not sure if it still works but you can use it as a base.
